File: js/libs/bitpack.js
function BitPack(data, b64) {
    if (b64 == undefined) {
        b64 = true;
    }

    if (b64) {
        try {
            this.data = atob(data);
        } catch (InvalidCharacterError) {
            this.data = '';
        }
    } else {
        this.data = data;
    }
}

BitPack.prototype.byteAt = function(index) {
    if (index >= this.len() || index < 0) {
        throw new RangeError();
    }
    return this.data.charCodeAt(index) & 0xff;
};

BitPack.prototype.setByte = function(index, value) {
    if (index >= this.len() || index < 0) {
        throw new RangeError();
    }
    this.data = this.data.substr(0, index) + String.fromCharCode(value) +
        this.data.substr(index + 1);    
};

BitPack.prototype.bitAt = function(index, bitIndex) {
    if (bitIndex == undefined) {
        bitIndex = index % 8;
        index = Math.floor(index / 8);
    }

    if (this.byteAt(index) & (0x80 >> bitIndex)) {
        return true;
    }

    return false;
};

BitPack.prototype.setBit = function(index, bitIndex, value) {
    if (value == undefined) {
        value = bitIndex;
        bitIndex = index % 8;
        index = Math.floor(index / 8);
    }

    var byte = this.byteAt(index);
    if (value) {
        byte |= 0x80 >> bitIndex;
    } else {
        byte ^= 0x80 >> bitIndex;
    }

    this.setByte(index, byte);
};

BitPack.prototype.len = function() {
    return this.data.length;
};

BitPack.prototype.or = function(pack, index) {
    index = index | 0;

    // if this reads past the end of our data, throw a range error
    if (pack.len() > index + this.len()) {
        throw new RangeError();
    }

    var resultData = '';
    for (var i = 0; i < pack.len(); i++) {
        resultData += String.fromCharCode(
            this.byteAt(index + i) | pack.byteAt(i)
        );
    }

    return new BitPack(resultData, false);
};

BitPack.prototype.boolOr = function(pack, index) {
    index = index | 0;

    // oring with an empty pack should always return true.
    if (pack.len() == 0) {
        return true;
    }

    // if this check reads past the end of our data, return false
    if (index + pack.len() > this.len()) {
        return false;
    }

    
    for (var i = 0; i < pack.len(); i++) {
        if (this.byteAt(index + i) | pack.byteAt(i)) {
            return true;
        }
    }
    return false;
};

BitPack.prototype.and = function(pack, index) {
    index = index | 0;

    // if this reads past the end of our data, throw a range error
    if (pack.len() > index + this.len()) {
        throw new RangeError();
    }

    var resultData = '';
    for (var i = 0; i < pack.len(); i++) {
        resultData += String.fromCharCode(
            this.byteAt(index + i) & pack.byteAt(i)
        );
    }

    return new BitPack(resultData, false);
};

BitPack.prototype.boolAnd = function(pack, index) {
    index = index | 0;

    // anding with an empty pack should always return true.
    if (pack.len() == 0) {
        return true;
    }

    // if this reads past the end of our data, throw a range error
    if (index + pack.len() > this.len()) {
        throw new RangeError();
    }
    
    for (var i = 0; i < pack.len(); i++) {
        if (this.byteAt(index + i) & pack.byteAt(i)) {
            return true;
        }
    }
    return false;
};

BitPack.prototype.slice = function(start, end) {
    return new BitPack(this.data.slice(start, end), false);
};

File: python/bitpack.py
# WARNING: This is a customized version of bitpack.py and should not be removed
# from this directory until the dependency on this extra functionality is no
# longer needed or the functionality exists upstream.
# This file adds:
#   BitPack.set_bit()
#   BitPack.set_byte()
import base64


class BitPack():

    def __init__(self, data='', b64=True):
        if isinstance(data, bytes):
            self.data = data
        else:
            self.data = data.encode('raw_unicode_escape')

        if b64:
            self.data = base64.b64decode(self.data)

    def base64(self):
        return base64.b64encode(self.data)

    def byte_at(self, index):
        return self.data[index]

    def bit_at(self, index, bit_index=None):
        if bit_index is None:
            bit_index = index % 8
            index = index // 8

        if self.byte_at(index) & (0x80 >> bit_index):
            return True

        return False

    def __len__(self):
        return len(self.data)

    def bit_or(self, pack, index=0):
        # if this reads past the end of our data, raise an index error
        if len(pack) > index + len(self):
            raise IndexError('Cannot or beyond end of pack')

        result = bytes()
        for i in range(0, len(pack)):
            result += bytes(chr(self.byte_at(index + i) | pack.byte_at(i)),
                            'raw_unicode_escape')

        return BitPack(result, b64=False)

    def __or__(self, pack):
        return self.bit_or(pack)

    def bool_or(self, pack, index=0):
        # oring with an empty pack should always return true
        if len(pack) == 0:
            return True

        # if this check reads past the end of our data, return false
        if index + len(pack) > len(self):
            return False

        for i in range(0, len(pack)):
            if self.byte_at(index + i) | pack.byte_at(i):
                return True

        return False

    def bit_and(self, pack, index=0):
        # if this reads past the end of our data, raise an index error
        if len(pack) > index + len(self):
            raise IndexError('Cannot or beyond end of pack')

        result = bytes()
        for i in range(0, len(pack)):
            result += bytes(chr(self.byte_at(index + i) & pack.byte_at(i)),
                            'raw_unicode_escape')

        return BitPack(result, b64=False)

    def __and__(self, pack):
        return self.bit_and(pack)

    def bool_and(self, pack, index=0):
        # anding with an empty pack should always return true
        if len(pack) == 0:
            return True

        # if this reads past the end of our data, raise an index error
        if index + len(pack) > len(self):
            raise IndexError('Cannot or beyond end of pack')

        for i in range(0, len(pack)):
            if self.byte_at(index + i) & pack.byte_at(i):
                return True

    def __getitem__(self, index):
        return BitPack(self.data[index], b64=False)

    def set_bit(self, value, index, bit_index=None):
        if bit_index is None:
            bit_index = index % 8
            index = index // 8

        byte = self.byte_at(index)

        if value:
            byte |= 0x80 >> bit_index
        elif byte & (0x80 >> bit_index):
            byte ^= 0x80 >> bit_index

        self.set_byte(index, byte)

    def set_byte(self, index, value):
        if index >= len(self) or index < 0:
            raise IndexError('Cannot index outside of pack bounds')

        self.data = self.data[0:index] + bytes([value]) + self.data[index+1:]
